src/crimson/osd/ops_executer.cc

```diff
@@ -100,51 +100,219 @@ seastar::future<> OpsExecuter::do_op_call(OSDOp& osd_op)
         }
       };
+
+      auto range = objects | boost::adaptors::filtered(in_my_namespace)
+                           | boost::adaptors::transformed(to_pglsed);
+      logger().debug("do_pgnls_common: finishing the 1st stage of pgls");
+      return seastar::when_all_succeed(std::begin(range),
+                                       std::end(range)).then(
+        [next=std::move(next)] (auto items) mutable {
+          // the sole purpose of this chaining is to pass `next` to 2nd
+          // stage altogether with items
+          logger().debug("do_pgnls_common: 1st done");
+          return seastar::make_ready_future<decltype(items), decltype(next)>(
+            std::move(items), std::move(next));
+      });
+  }).then(
+    [pg_end, filter] (const std::vector<std::tuple<bool,
+                                hobject_t>>& items, auto next) {
+      auto is_matched = [] (const auto& item) {
+        return std::get<bool>(item);
+      };
+      auto to_entry = [] (const auto& item) {
+        const auto& obj = std::get<hobject_t>(item);
+        return librados::ListObjectImpl{
+          obj.get_namespace(), obj.oid.name, obj.get_key()
+        };
+      };
+
       pg_nls_response_t response;
-      boost::copy(objects |
-        boost::adaptors::filtered(in_my_namespace) |
-        boost::adaptors::transformed([](const hobject_t& o) {
-          return librados::ListObjectImpl{o.get_namespace(),
-                                          o.oid.name,
-                                          o.get_key()}; }),
-        std::back_inserter(response.entries));
+      boost::push_back(response.entries,
+                       items | boost::adaptors::filtered(is_matched)
+                             | boost::adaptors::transformed(to_entry));
       response.handle = next.is_max() ? pg_end : next;
-      bufferlist bl;
-      encode(response, bl);
-      return seastar::make_ready_future<bufferlist>(std::move(bl));
+      ceph::bufferlist out;
+      encode(response, out);
+      logger().debug("{}: response.entries.size()=",
+                     __func__, response.entries.size());
+      return seastar::make_ready_future<ceph::bufferlist>(std::move(out));
+  });
+}
+
+seastar::future<> OpsExecuter::do_pgnls(OSDOp& osd_op)
+{
+  hobject_t lower_bound;
+  try {
+    ceph::decode(lower_bound, osd_op.indata);
+  } catch (const buffer::error& e) {
+    throw std::invalid_argument("unable to decode PGNLS handle");
+  }
+  return do_pgnls_common(lower_bound, os->oi.soid.get_namespace(), osd_op.op.pgls.count)
+    .then([&osd_op](bufferlist bl) {
+      osd_op.outdata = std::move(bl);
+      return seastar::now();
+  });
+}
+
+seastar::future<> OpsExecuter::do_pgnls_filtered(OSDOp& osd_op)
+{
+  std::string cname, mname, type;
+  auto bp = osd_op.indata.cbegin();
+  try {
+    ceph::decode(cname, bp);
+    ceph::decode(mname, bp);
+    ceph::decode(type, bp);
+  } catch (const buffer::error& e) {
```

Comment:
remove `e` as it's not used.

Reply:
Dropped from `do_pgnls()` and `do_pgnls_filtered()` as well.